File: VB/Comparers.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports DevExpress.XtraScheduler

Namespace SchedulerSortResources

    Public MustInherit Class ResourceBaseComparer
        Inherits IComparer(Of Resource)
        Implements IComparer

#Region "IComparer Members"
        Private Function Compare(ByVal x As Object, ByVal y As Object) As Integer
            Return Me.CompareCore(x, y)
        End Function

        Public Function Compare(ByVal x As Resource, ByVal y As Resource) As Integer
            Return CompareCore(x, y)
        End Function

#End Region
        Protected Overridable Function CompareCore(ByVal x As Object, ByVal y As Object) As Integer
            Dim xRes As Resource = CType(x, Resource)
            Dim yRes As Resource = CType(y, Resource)
            If xRes Is Nothing OrElse yRes Is Nothing Then Return 0
            If Resource.Empty.Equals(xRes) OrElse Resource.Empty.Equals(yRes) Then Return 0
            Return Me.CompareResources(xRes, yRes)
        End Function

        Protected MustOverride Function CompareResources(ByVal xRes As Resource, ByVal yRes As Resource) As Integer
    End Class

    Public Class ResourceCaptionComparer
        Inherits ResourceBaseComparer

        Protected Overrides Function CompareResources(ByVal xRes As Resource, ByVal yRes As Resource) As Integer
            Return [String].Compare(xRes.Caption, yRes.Caption)
        End Function
    End Class

    Public Class ResourceCaptionReverseComparer
        Inherits ResourceBaseComparer

        Protected Overrides Function CompareResources(ByVal xRes As Resource, ByVal yRes As Resource) As Integer
            Return [String].Compare(yRes.Caption, xRes.Caption)
        End Function
    End Class

    Public Class ResourceNOfAppointmentsComparer
        Inherits ResourceBaseComparer

        Private schedulerStorage As SchedulerStorage

        Public Sub New(ByVal schedulerStorage As SchedulerStorage)
            Me.schedulerStorage = schedulerStorage
        End Sub

        Protected Overrides Function CompareResources(ByVal xRes As Resource, ByVal yRes As Resource) As Integer
            ' "Unassigned" resource
            If Convert.ToInt32(yRes.Id) Is -1 Then Return 1
            ' "Unassigned" resource
            If Convert.ToInt32(xRes.Id) Is -1 Then Return -1
            Dim order As Integer = schedulerStorage.Appointments.Items.FindAll(Function(e) e.ResourceId.Equals(yRes.Id)).Count - schedulerStorage.Appointments.Items.FindAll(Function(e) e.ResourceId.Equals(xRes.Id)).Count
            If order Is 0 Then Return schedulerStorage.Resources.Items.IndexOf(xRes) - schedulerStorage.Resources.Items.IndexOf(yRes)
            Return order
        End Function
    End Class
End Namespace

File: CS/Comparers.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DevExpress.XtraScheduler;

namespace SchedulerSortResources {
    public abstract class ResourceBaseComparer : IComparer<Resource>, IComparer {
        #region IComparer Members
        int IComparer.Compare(object x, object y) {
            return CompareCore(x, y);
        }
        public int Compare(Resource x, Resource y) {
            return CompareCore(x, y);
        }
        #endregion

        protected virtual int CompareCore(object x, object y) {
            Resource xRes = (Resource)x;
            Resource yRes = (Resource)y;
            
            if (xRes == null || yRes == null)
                return 0;
            if (Resource.Empty.Equals(xRes) || Resource.Empty.Equals(yRes))
                return 0;
            
            return CompareResources(xRes, yRes);
        }

        protected abstract int CompareResources(Resource xRes, Resource yRes);
    }

    public class ResourceCaptionComparer : ResourceBaseComparer {
        protected override int CompareResources(Resource xRes, Resource yRes) {
            return String.Compare(xRes.Caption, yRes.Caption);
        }
    }

    public class ResourceCaptionReverseComparer : ResourceBaseComparer {
        protected override int CompareResources(Resource xRes, Resource yRes) {
            return String.Compare(yRes.Caption, xRes.Caption);
        }
    }

    public class ResourceNOfAppointmentsComparer : ResourceBaseComparer {
        private SchedulerStorage schedulerStorage;

        public ResourceNOfAppointmentsComparer(SchedulerStorage schedulerStorage) {
            this.schedulerStorage = schedulerStorage;
        }

        protected override int CompareResources(Resource xRes, Resource yRes) {
            // "Unassigned" resource
            if (Convert.ToInt32(yRes.Id) == -1)
                return 1;
            
            // "Unassigned" resource
            if (Convert.ToInt32(xRes.Id) == -1)
                return -1;

            int order = schedulerStorage.Appointments.Items.FindAll(e => e.ResourceId.Equals(yRes.Id)).Count -
                schedulerStorage.Appointments.Items.FindAll(e => e.ResourceId.Equals(xRes.Id)).Count;

            if (order == 0)
                return schedulerStorage.Resources.Items.IndexOf(xRes) - schedulerStorage.Resources.Items.IndexOf(yRes);

            return order;
        }
    }
}
